zapcore/sampler.go

```diff
@@ -42,8 +42,13 @@ func newCounters() *counters {
 	return &counters{}
 }
 
+var nullCounter counter
+
 func (cs *counters) get(lvl Level, key string) *counter {
 	i := lvl - _minLevel
+	if i < 0 {
+		return &nullCounter
```

Comment:
The behaviour here is that all unknown level / any message will be sampled as if they were a single level + message, which I don't think is the intended behaviour. It will likely lead to over-sampling

Looking at the implementation, a simple alternative could be to min/max `i` so it's within the bounds:
```
if i < 0 {
  i = 0
}
if i >= len(cs) {
  i = len(cs) - 1
}
```

Alternatively, we could have a separate set of counters for "unknown level", but I think the above is a little bit simpler.

Reply:
I'll admit I don't know enough about zap to say what the "right" answer is or even exactly what's being sampled.  I assumed that the counts for out-of-bounds level just weren't interesting (since they don't have names).

If the desire is just to saturate at the lowest/highest named levels then your solution works.  But that seems like it might confuse users?

If you want to actually count per-level, then maybe we want a `map[int][_countersPerLevel]counter`, but then we need some atomic accessor.  To avoid that in most cases, you could keep the array and fall back to the map only if the level is out-of-bounds?  E.g.

```
    // ...
    if i >= 0 && i < len(cs.named) {
        return cs.named[i][j]
    }
    return cs.getUnnamed(i, j)
}

func (cs *counters) getUnnamed(i, j int) *counter {
    cs.mu.Lock()
    defer cs.mu.Unlock()
    if cs.unnamed[i] == nil {
        cs.unnamed[i] = [_countersPerLevel]counter{}
    }
    return &cs.unnamed[i][j]
}
```

Maybe simpler is my original plan - just add some fixed number of counters at one or both ends and document it.  Your `counters` type gets bigger either way, but I am not confident to say if that matters or not.

I'll approach it however you think is best - my need is small, but I don't want to have disproportionate impact.

Reply:
My alternative proposal was mostly to localize this change to where we use the level to convert to an index, but I agree that affecting other levels may be surprising.

I think the preferred behaviour is for the sampler to ignore unknown levels. Instead of making any change here, we can instead do the level check at the caller,
https://github.com/uber-go/zap/blob/v1.18.1/zapcore/sampler.go#L200

If the level is unsupported, then we "fail open" by calling the underlying core's `Check`.